Framework/DataHandling/src/LoadSpec.cpp

```diff
@@ -50,109 +50,132 @@ void LoadSpec::init() {
-        j++;
-        spectra[working_with_spectrum_nbr].dataY().push_back(input[j]);
-        j++;
-        spectra[working_with_spectrum_nbr].dataE().push_back(input[j]);
-        nBins = j / 3;
-      }
+    if (static_cast<size_t>(specNum) == (nSpectra - 1)) {
+      auto histogram = localWorkspace->histogram(specNum);
+      readHistogram(input, histogram);
+      localWorkspace->setHistogram(specNum, histogram);
     }
   } catch (...) {
   }
 
-  try {
-    int nSpectra = spectra_nbr;
-    MatrixWorkspace_sptr localWorkspace =
-        boost::dynamic_pointer_cast<MatrixWorkspace>(
-            WorkspaceFactory::Instance().create("Workspace2D", nSpectra, nBins,
-                                                nBins));
-
-    localWorkspace->getAxis(0)->unit() =
-        UnitFactory::Instance().create(getProperty("Unit"));
-
-    for (int i = 0; i < nSpectra; i++) {
-      localWorkspace->dataX(i) = spectra[i].dataX();
-      localWorkspace->dataY(i) = spectra[i].dataY();
-      localWorkspace->dataE(i) = spectra[i].dataE();
-      // Just have spectrum number start at 1 and count up
-      localWorkspace->getSpectrum(i).setSpectrumNo(i + 1);
+  setProperty("OutputWorkspace", localWorkspace);
+}
+
+/**
+ * Read the total number of specrta contained in the file
+ *
+ * @param file :: file stream to read from
+ * @return a size_t representing the number of spectra
+ */
+size_t LoadSpec::readNumberOfSpectra(std::ifstream &file) const {
+  // determine the number of lines starting by #L
+  // as there is one per set of data
+  size_t spectra_nbr = 0;
+  std::string str;
+  while (getline(file, str)) {
+    if (str.empty())
+      continue;
+    if (str[0] == '#' && str[1] == 'L') {
+      spectra_nbr++;
     }
+  }
+  return spectra_nbr;
+}
 
-    setProperty("OutputWorkspace", localWorkspace);
-  } catch (Exception::NotFoundError &) {
-    // Asked for dimensionless workspace (obviously not in unit factory)
+/**
+ * Read a single line from the file into the data buffer
+ *
+ * @param line :: the current line in the file to process
+ * @param buffer :: the buffer to store loaded data in
+ */
+void LoadSpec::readLine(const std::string &line,
+                        std::vector<double> &buffer) const {
+  if (!line.empty() && line[0] != '#') {
+    typedef Mantid::Kernel::StringTokenizer tokenizer;
+    const std::string sep = " ";
+    tokenizer tok(line, sep, Mantid::Kernel::StringTokenizer::TOK_IGNORE_EMPTY);
+    for (const auto &beg : tok) {
+      buffer.push_back(std::stod(beg));
+    }
   }
 }
 
+/**
+ * Convert the data from the input buffer to a histogram object
+ *
+ * @param input :: the input buffer containing the raw data
+ * @param histogram :: the histogram object to fill with values
+ * */
+void LoadSpec::readHistogram(const std::vector<double> &input,
+                             HistogramData::Histogram &histogram) const {
+  std::vector<double> x, y, e;
+
+  size_t index = 0;
+  for (; index < input.size() - 1; index++) {
```

Comment:
I think this was just based on how it worked at line 121 in the original file. But you're right we should probably not make assumptions when we can calculate a proper upper limit from the size of `input`.